### ISSUE TITLE
Help needed with FFT library for OpenCL

### ISSUE THREAD

**Issue by craabreu** [openmm/openmm#3710]
I'm writing an OpenMM plugin that requires batched FFT calculations. I decided to use [VkFFT](https://github.com/DTolm/VkFFT) because it works with both CUDA and OpenCL (and it is slightly faster than `cuFFT` in Nvidia GPUs). In OpenCL, the FFT planning involves passing the device, context, and memory buffers as pointers of types `cl_device_id*`, `cl_context*`, and `cl_mem*`, respectively. I am trying to access these objects by doing the following:
```c++
    cl::Device clDevice = context.getDevice();
    cl::Context clContext = context.getContext();
    cl::Buffer clInputBuffer = grid1.getDeviceBuffer();
    cl::Buffer clOutputBuffer = grid2.getDeviceBuffer();

    cl_device_id device = clDevice.get();
    cl_context ctxt = clContext.get();
    cl_mem inputBuffer = clInputBuffer.get();
    cl_mem outputBuffer = clOutputBuffer.get();
```
Here, `context` is an `OpenCLContext`, while `grid[1|2]` are `OpenCLArray`s. The `.get()` method of OpenCL C++ wrappers is documented [here](http://github.khronos.org/OpenCL-CLHPP/classcl_1_1detail_1_1_wrapper.html#adc64140f053332174a28ea24db94842c).

Then, I pass `&device`, `&ctxt`, `&inputBuffer`, and `&outputBuffer` to VkFFT:
```c++
    VkFFTConfiguration config = {};
    config.device = &device;
    config.context = &ctxt;
    config.inputBuffer = &inputBuffer;
    config.buffer = &outputBuffer;
```
However, this procedure causes a segfault. I have also tried using the `()` operator instead of `.get()`, as documented [here](http://github.khronos.org/OpenCL-CLHPP/classcl_1_1detail_1_1_wrapper.html#ac62b1a8325e7c06fd2cb125a1730bf34), but it failed too.

More details, including a successful test code for CUDA and the unsuccessful one for OpenCL, are available in this external [PR](https://github.com/craabreu/openmm/pull/2/files).

I'd really appreciate it if someone could help me figure out the cause of the issue. The plugin I am working on aims at splitting PME calculations by groups of particles, but I believe that VkFFT could also be used to improve the efficiency of OpenMM's `NonbondedForce` in the OpenCL platform.

**Comment by peastman** [comment]
Can you try running it inside `gdb` or `lldb`?  Run it until it segfaults, then type `bt` to get a stack trace showing where it happened.  If you compile in debug mode, you'll get a more detailed stack trace.  This will hopefully give a clue where to start debugging.

Make sure `grid1` and `grid2` have been initialized.  If they haven't been, that would cause a crash.

**Comment by craabreu** [comment]
Thank you, @peastman.

I was able to solve the problem. Unlike what I suspected, it was unrelated to OpenCL's C++ wrapper. I was unaware of a subtle difference between the CUDA and OpenCL versions of VkFFT. In the former, we must define the CUDA stream during the FFT planning stage, while in the latter, we must pass the OpenCL command queue at every execution. I don't know why they had to make it this way.
